gateway/main.go

```diff
@@ -14,48 +18,136 @@ const (
+type GQLRequest struct {
+	q          string
+	vars       map[string]interface{}
+	header     http.Header
+	httpClient *http.Client
+}
+
+type GQLError struct {
+	Message string `json:"message"`
+}
+
+func (e GQLError) Error() string {
+	return e.Message
+}
+
+type GQLResponse struct {
+	Errors []GQLError  `json:"errors"`
+	Data   interface{} `json:"data"`
+}
+
+func (g *Gateway) authorize(header http.Header) error {
+	header.Add("x-source", CLI_SOURCE_HEADER)
+
+	if g.cfg.RailwayProductionToken != "" {
+		header.Add("project-access-token", g.cfg.RailwayProductionToken)
+	} else {
+		user, err := g.cfg.GetUserConfigs()
+		if err != nil {
+			return err
+		}
+		header.Add("authorization", fmt.Sprintf("Bearer %s", user.Token))
+	}
 
 	return nil
 }
 
-func GetHost() string {
-	baseURL := "https://backboard.railway.app"
-	if configs.IsDevMode() {
-		baseURL = "http://localhost:8082"
+func (g *Gateway) NewRequestWithoutAuth(query string) *GQLRequest {
+	return &GQLRequest{
+		q:          query,
+		header:     http.Header{},
+		httpClient: g.httpClient,
+		vars:       make(map[string]interface{}),
 	}
-	return baseURL
 }
 
-func GetGQLHost() string {
-	baseURL := GetHost()
-	return fmt.Sprintf("%s/graphql", baseURL)
+func (g *Gateway) NewRequestWithAuth(query string) (*GQLRequest, error) {
+	gqlReq := g.NewRequestWithoutAuth(query)
+
+	err := g.authorize(gqlReq.header)
+	if err != nil {
+		return gqlReq, err
+	}
+
+	return gqlReq, nil
 }
 
-func New() *Gateway {
-	gqlClient := gql.NewClient(GetGQLHost())
-	gqlClient.Log = func(s string) {
-		// Uncomment this for verbose query logging
-		// fmt.Println(s)
+func (r *GQLRequest) Run(ctx context.Context, resp interface{}) error {
+	var requestBody bytes.Buffer
+	requestBodyObj := struct {
+		Query     string                 `json:"query"`
+		Variables map[string]interface{} `json:"variables"`
+	}{
+		Query:     r.q,
+		Variables: r.vars,
 	}
-	return &Gateway{
-		cfg:       configs.New(),
-		gqlClient: gqlClient,
+	if err := json.NewEncoder(&requestBody).Encode(requestBodyObj); err != nil {
+		return errors.Wrap(err, "encode body")
+	}
+
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/graphql", GetHost()), &requestBody)
+	if err != nil {
+		return err
+	}
+
+	req = req.WithContext(ctx)
+	req.Header = r.header
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json; charset=utf-8")
+	res, err := r.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, res.Body); err != nil {
+		return err
+	}
+
+	// TODO: Handle auth errors and other things in a special way
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return fmt.Errorf("Response not successful status=%d", res.StatusCode)
 	}
+
+	gr := &GQLResponse{
+		Data: resp,
+	}
+	if err := json.NewDecoder(&buf).Decode(&gr); err != nil {
+		return errors.Wrap(err, "decoding response")
+	}
+	if len(gr.Errors) > 0 {
```

Comment:
Do we want to only return one? Wondering if there's a case here to return all of em joined

Reply:
Could? I just pulled this from the old client we were using. Any suggestions welcome

Reply:
Oh very fair

Ya I'd probably err on the side of doing a join and returning one bigass error, just so we don't bury context

Reply:
Hmm, we never actually show these errors anywhere right now but I updated it for later